File: Python/sendNotificationToAPNS.py
import json
import time

import httpx
import jwt

Key_ID = ""
Team_ID = ""
APNS_TOPIC = ""
APNS_URL_DEV = ""

Device_Token = "put your device Token here"


def Init():
    # 从config.json打开文件
    with open("config.json", "r") as file:
        config = json.load(file)
        global Key_ID, Team_ID, APNS_TOPIC, APNS_URL_DEV
        Key_ID = config["Key_ID"]
        Team_ID = config["Team_ID"]
        APNS_TOPIC = config["APNS_TOPIC"]
        APNS_URL_DEV = config["APNS_URL_DEV"]


def generate_token():
    key = ""
    with open("AuthKey.p8", "r") as file:
        key = file.read()
    header = {"alg": "ES256", "kid": Key_ID}
    payload = {"iss": Team_ID, "iat": int(time.time())}
    bash64_signature = jwt.encode(
        key=key,
        headers=header,
        algorithm="ES256",
        payload=payload,
    )
    return bash64_signature


def send_notification(token):
    headers = {
        "apns-topic": APNS_TOPIC,
        "apns-priority": "5",
        "apns-push-type": "alert",
        "authorization": "bearer " + token,
        "host": APNS_URL_DEV,
        "apns-expiration": "0",
    }
    payload = {
        "aps": {
            "alert": {
                "title": "Title",
                "subtitle": "Subtitle",
                "body": " Body",
                "action": "PLAY",
            },
            "badge": 5,
        }
    }
    url = APNS_URL_DEV + "/3/device/" + Device_Token

    with httpx.Client(http2=True) as client:
        resp = client.post(url, headers=headers, json=payload)
        print(resp.text)


if __name__ == "__main__":
    Init()
    token = generate_token()
    send_notification(token)

# """
# curl -v \
#   --header "authorization: bearer eyJhbzxcxssFUzI1NiIsImtpZCI6Ikw2S1JQQ0I5QjIiLCJ0eXAiOiJKV1QifQ.eyJpc3MiOiIzVE45UDJZQjgyIiwiaWF0IjoxNzEyMjMwNDEyfQ.SFr2hQOtI9zoXpcdL2SoZW5yZzpvaxVai8-8FdxrylNqDmUofryEZI_ogdeiMoY4Si3-3vMeL0nFqX8O6gafVw" \
#   --header "apns-topic: com.ths-express.stdexpress" \
#   --header "apns-push-type: alert" \
#   --header "apns-priority: 5" \
#   --header "apns-expiration: 0" \
#   --data '{"aps":{"alert":{"title":"title","subtitle":"subtitle","body":"body"}}}' \
#   --http2  https://api.development.push.apple.com:443/3/device/ced1949617ee5c2f826eq51072bf5726b1ed8bdfb09e851631427eb0487f59d5
# """

File: Java/com/src/com/qzh/sendNotificationToAPNS.java
package com.qzh;

import com.qzh.request_url;
import org.json.JSONObject;
import org.json.JSONException;
import java.io.BufferedInputStream;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Base64;
import java.util.Properties;

import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.JwtBuilder;
import io.jsonwebtoken.Jwts.SIG;

import java.security.KeyFactory;
import java.security.KeyStoreException;
import java.security.NoSuchAlgorithmException;
import java.security.UnrecoverableEntryException;
import java.security.cert.CertificateException;
import java.security.interfaces.ECPrivateKey;
import java.security.spec.InvalidKeySpecException;
import java.security.spec.PKCS8EncodedKeySpec;

public class sendNotificationToAPNS {
    private static String APNS_CERTIFICATE_PATH;
    private static String APNS_TOPIC;
    private static String Key_ID;
    private static String Team_ID;
    private static String token = "";
    private static String APNS_URL = "";
    private static String APNS_URL_PROD = "";
    private static String APNS_URL_DEV = "";

    public sendNotificationToAPNS()
            throws KeyStoreException, NoSuchAlgorithmException, CertificateException,
            UnrecoverableEntryException, IOException, InvalidKeySpecException {
        Properties prop = new Properties();
        try {
            InputStream in = new BufferedInputStream(
                    new FileInputStream("/root/code/StdExpress/com/src/com/qzh/apple.properties"));
            prop.load(in);

            APNS_CERTIFICATE_PATH = prop.getProperty("APNS_CERTIFICATE_PATH");
            APNS_TOPIC = prop.getProperty("APNS_TOPIC");
            Key_ID = prop.getProperty("Key_ID");
            Team_ID = prop.getProperty("Team_ID");
            APNS_URL_PROD = prop.getProperty("APNS_URL_PROD");
            APNS_URL_DEV = prop.getProperty("APNS_URL_DEV");
            APNS_URL = APNS_URL_PROD;

            token = generateToken();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public sendNotificationToAPNS setDev() {
        APNS_URL = APNS_URL_DEV;
        return this;
    }

    public boolean sendNotification(String deviceToken, String title, String subtitle, String body, Integer badge) {
        // payload = {"aps": {"alert": {"title": str(int(time.time())), "subtitle":
        // "Subtitle", "body": " Body","action": "PLAY",},"badge": 50,}}
        try {

            JSONObject payload = payload(title, subtitle, body, badge);

            String url = APNS_URL + "/3/device/" + deviceToken;
            System.out.println(url);

            JSONObject headers = new JSONObject();
            headers.put("apns-topic", APNS_TOPIC);
            headers.put("authorization", "bearer " + token);
            headers.put("apns-push-type", "alert");
            headers.put("apns-priority", "5");

            String resp = request_url.sendPost(url, headers, payload.toString());
            System.out.println(resp);

        } catch (Exception e) {
            e.printStackTrace();
            return false;
        }
        return true;
    }

    private JSONObject payload(String title, String subtitle, String body, Integer badge) {
        try {

            JSONObject payload = new JSONObject();
            JSONObject aps = new JSONObject();
            JSONObject alert = new JSONObject();
            alert.put("title", title);

            if (subtitle != null) {
                alert.put("subtitle", subtitle);
            }

            alert.put("body", body);

            aps.put("alert", alert);
            if (badge != null) {
                aps.put("badge", badge);
            }
            payload.put("aps", aps);

            return payload;

        } catch (JSONException e) {
            e.printStackTrace();
            return null;
        }
    }

    private String generateToken()
            throws KeyStoreException, NoSuchAlgorithmException, CertificateException, IOException,
            UnrecoverableEntryException, InvalidKeySpecException {
        String signature = "";

        byte[] keyBytes = Files.readAllBytes(Paths.get(APNS_CERTIFICATE_PATH));
        String privateKeyContent = new String(keyBytes).replaceAll("\\n", "")
                .replace("-----BEGIN PRIVATE KEY-----", "").replace("-----END PRIVATE KEY-----", "");
        KeyFactory kf = KeyFactory.getInstance("EC");
        PKCS8EncodedKeySpec keySpec = new PKCS8EncodedKeySpec(Base64.getDecoder().decode(privateKeyContent));
        ECPrivateKey privateKey = (ECPrivateKey) kf.generatePrivate(keySpec);

        JwtBuilder jwtbuilder = Jwts.builder();
        signature = jwtbuilder
                .header()
                .add("kid", Key_ID)
                .add("typ", "JWT")
                .add("alg", "ES256")
                .and()
                .claims()
                .add("iss", Team_ID)
                .add("iat", System.currentTimeMillis() / 1000)
                .and()
                .signWith(privateKey, SIG.ES256)
                .compact();

        return signature;
    }

}
